lib/cusparse/linalg.jl

```diff
@@ -48,6 +48,66 @@ function sum_dim2(A::CuSparseMatrixCSR{T}) where {T}
     return colsum
 end
 
+function compute_threads(max_threads, wanted_threads, shuffle, dev)
+    if wanted_threads > max_threads
+        shuffle ? prevwarp(dev, max_threads) : max_threads
+    else
+        wanted_threads
+    end
+end
+
+function kernel_dot_csc(y::CuDeviceVector{T1}, colPtr::CuDeviceVector{T2}, rowVal::CuDeviceVector{T2},
+    nzVal::CuDeviceVector{T1}, x::CuDeviceVector{T1}, result::CuDeviceVector{T1}, n::Integer, shuffle) where {T1,T2}
+
+    thread_idx = threadIdx().x
+    index = (blockIdx().x-1) * blockDim().x + thread_idx
+    stride = blockDim().x * gridDim().x
+
+    tmp = zero(T1)
+    if index <=n
+        @inbounds for col in index:stride:n
+            for j in (colPtr[col]):(colPtr[col+1]-1)
+                row = rowVal[j]
+                val = nzVal[j]
+                tmp += dot(y[row], val, x[col])
+            end
+        end
+    end
+
+    reduced_val = CUDA.reduce_block(+, tmp, zero(T1), shuffle)
+
+    if thread_idx == 1
+        @inbounds result[blockIdx().x] = reduced_val
+    end
+    return
+end
```

Comment:
Since these are single-use kernels, I'd inline them as simply `function kernel` inside the calling function.